### ISSUE TITLE
pkg/base64: Golang base64 decoding bug can lead to PutObject/PutObjectPart corruption on server side.

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio#1608]
To reproduce first configure `aws cli`

```
$ aws --endpoint-url http://localhost:9000 s3api put-object --bucket testbucket --key testfile --body testfile --content-md5 WvLTlMrX9NpYDQlEIFlnDA==
{
    "ETag": "\"5af2d394cad7f4da580d09442059670c\""
}

$ md5 testfile
MD5 (testfile) = 5af2d394cad7f4da580d09442059670c
```

Success first attempt valid base64 of the file. 

```
$ aws --endpoint-url http://localhost:9000 s3api put-object --bucket testbucket --key testfile --body testfile --content-md5 WvLTlMrX9NpYDQlEIFlnDB==
{
    "ETag": "\"5af2d394cad7f4da580d09442059670c\""
}
$ md5 testfile
MD5 (testfile) = 5af2d394cad7f4da580d09442059670c
```

Success second attempt for invalid base64 of the file. 

```
$ aws --endpoint-url http://localhost:9000 s3api put-object --bucket testbucket --key testfile --body testfile --content-md5 WvLTlMrX9NpYDQlEIFlnDw==

A client error (BadDigest) occurred when calling the PutObject operation: The Content-Md5 you specified did not match what we received.
```

Change any other character this shows proper error message.

**Comment by harshavardhana** [comment]
https://github.com/golang/go/issues/15656 - for more information.

**Comment by krisis** [comment]
@harshavardhana The fact the two different base64 encoded values could be decoded to the same MD5 doesn't affect the correctness of our PutObject/PutObjectPart implementation. To prove that, we need to show that md5sum of a different object cannot be equal to any of the 15 other non-canonical base64 representations. 

For a given md5sum (of 128-bit), we would have 5 \* 24-bit groups and one 8-bit residue. This 8-bits are padded with 16 more bits to form the last 24-bit group as per the RFC. For canonical representation, the RFC mandates that padding bits are zero'd.

In md5sum's case, 8-bits are split into 6 and remaining 2-bits with 4 padding. The 4 padding bits allow 16 different representations including the canonical one, with all zeros. Note that these representations are not part of the original bits from md5sum and therefore a different md5sum cannot be equal to one of the non-canonical representations.

In conclusion, we don't need strict base64 decoding for proper functioning of minio.

**Comment by harshavardhana** [comment]
> In conclusion, we don't need strict base64 decoding for proper functioning of minio.

Ah makes sense.

**Comment by harshavardhana** [comment]
Closing this as not an issue, relevant bug is available upstream Golang.

**Comment by harshavardhana** [comment]
https://go-review.googlesource.com/#/c/24964/ - relevant patch upstream FWIW!
